include/dali/imgcodec/image_decoder.h

```diff
@@ -0,0 +1,366 @@
+// Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_IMGCODEC_IMAGE_DECODER_H_
+#define DALI_IMGCODEC_IMAGE_DECODER_H_
+
+#include <condition_variable>
+#include <map>
+#include <memory>
+#include <mutex>
+#include <string>
+#include <typeinfo>
+#include <utility>
+#include <vector>
+#include "dali/imgcodec/image_decoder_interfaces.h"
+#include "dali/imgcodec/image_format.h"
+#include "dali/pipeline/data/tensor_vector.h"
+
+namespace dali {
+namespace imgcodec {
+
+inline bool is_cpu_decoder(ImageDecoderFactory *factory) {
+  const auto &properties = factory->GetProperties();
+  return !properties.gpu_output && !(properties.supported_input_kinds & InputKind::DeviceMemory);
+}
+
+class DLL_PUBLIC ImageDecoder : public ImageDecoderInstance, public ImageParser {
+ public:
+  ImageDecoder(int device_id,
+               bool lazy_init,
+               const std::map<std::string, any> &params = {})
+  : ImageDecoder(
+      device_id,
+      lazy_init,
+      params,
+      device_id == CPU_ONLY_DEVICE_ID ? is_cpu_decoder
+                                      : [](ImageDecoderFactory *) { return true; }) {}
+
+  /**
+   * @brief Construct a new ImageDecode, allowing the user to filter the decoders
+   *
+   * @param device_id       CUDA device ordinal or -1 for current device or CPU_ONLY_DEVICE_ID
+   * @param lazy_init       if true, the construction of sub-decoders is deferred until they are
+   *                        needed - this may be usefule when decoding datasets with
+   *                        few image formats
+   * @param params          parameters passed to the construction of the decoders;
+   *                        when using lazy initialization, the parameters can be set later
+   * @param decoder_filter  a predicate applied to decoder factories, which allows the caller
+   *                        to select a subset with desired properties
+   */
+  ImageDecoder(int device_id,
+               bool lazy_init,
+               const std::map<std::string, any> &params,
+               std::function<bool(ImageDecoderFactory *)> decoder_filter);
+
+  ~ImageDecoder();
+
+  bool CanParse(ImageSource *encoded) const override {
+    auto *f = FormatRegistry().GetImageFormat(encoded);
```

Comment:
why not move all implementations to the cc file?

Reply:
Done.